File: server/src/domain/quantity.ts
/**
 * Quantity as an integer scaled by 1000 (3 decimal places), so 0.5 kg = 500,
 * 3 pieces = 3000. Never a float — see PRD.md section 24, DECISIONS.md D019.
 *
 * Chosen over a plain integer because shops sell fractional quantities
 * (0.5 kg, 1.25 kg, 2.5 litre), not just whole pieces.
 */
export type Quantity = number & { readonly __brand: 'Quantity' }

const SCALE = 1000

export function quantity(scaledUnits: number): Quantity {
  if (!Number.isInteger(scaledUnits)) {
    throw new TypeError(
      `Quantity must be an integer (already scaled by ${SCALE}), got ${scaledUnits}. ` +
        'Quantity is never a float (D019).',
    )
  }
  return scaledUnits as Quantity
}

/** Converts a human-entered decimal amount (e.g. 0.5) into a Quantity. */
export function quantityFromDecimal(decimalAmount: number): Quantity {
  const scaled = Math.round(decimalAmount * SCALE)
  return quantity(scaled)
}

/** Converts a Quantity back to the decimal amount a person would read (e.g. 0.5). */
export function quantityToDecimal(qty: Quantity): number {
  return qty / SCALE
}

export const ZERO_QUANTITY: Quantity = quantity(0)

export function addQuantity(a: Quantity, b: Quantity): Quantity {
  return quantity(a + b)
}

export function subtractQuantity(a: Quantity, b: Quantity): Quantity {
  return quantity(a - b)
}

/** Sums a list of Quantity values (e.g. StockMovement.quantity_delta). Empty list sums to zero. */
export function sumQuantity(values: readonly Quantity[]): Quantity {
  return values.reduce(addQuantity, ZERO_QUANTITY)
}

File: android/app/src/main/java/com/hisab/app/domain/Quantity.kt
package com.hisab.app.domain

import kotlin.math.roundToLong

/**
 * Quantity as an integer scaled by 1000 (3 decimal places), so 0.5 kg = 500,
 * 3 pieces = 3000. Never a float — see PRD.md section 24, DECISIONS.md D019.
 *
 * Chosen over a plain integer because shops sell fractional quantities
 * (0.5 kg, 1.25 kg, 2.5 litre), not just whole pieces.
 */
private const val SCALE = 1000

@JvmInline
value class Quantity(
    val scaledUnits: Long,
) {
    operator fun plus(other: Quantity): Quantity = Quantity(scaledUnits + other.scaledUnits)

    operator fun minus(other: Quantity): Quantity = Quantity(scaledUnits - other.scaledUnits)

    /** The decimal amount a person would read, e.g. 0.5. */
    fun toDecimal(): Double = scaledUnits.toDouble() / SCALE

    companion object {
        val ZERO = Quantity(0)

        /** Converts a human-entered decimal amount (e.g. 0.5) into a Quantity. */
        fun fromDecimal(decimalAmount: Double): Quantity = Quantity((decimalAmount * SCALE).roundToLong())
    }
}

/** Sums a list of Quantity values (e.g. StockMovement.quantity_delta). Empty list sums to zero. */
fun List<Quantity>.sum(): Quantity = fold(Quantity.ZERO) { acc, value -> acc + value }
